lib/src/clojure_lsp/feature/file_management.clj

```diff
@@ -84,37 +85,56 @@
     (find-changed-elems-by usage-signature old-var-usages new-var-usages)))
 
 (defn reference-filenames [filename old-local-analysis new-local-analysis db]
-  (let [project-analysis (into {} (q/filter-project-analysis-xf db) (:analysis @db))
-        source-paths (settings/get db [:source-paths])
-        changed-var-definitions (find-changed-var-definitions old-local-analysis new-local-analysis)
-        references-filenames (->> changed-var-definitions
-                                  ;; TODO: switch from nested-loop join to hash join
-                                  (mapcat #(q/find-references project-analysis % false db))
-                                  (map :filename))
+  (let [changed-var-definitions (find-changed-var-definitions old-local-analysis new-local-analysis)
+        ;; TODO: can we easily remove usages of external namespaces (clojure.core, etc.) here?
```

Comment:
yes, these are all db analysis, we should probably q/filter-project-analysis on both for better performance?

Reply:
It's a little more complex than that.... We were talking about this before in relation to custom linting a single file. It's easy to exclude usages *from* an external ns, but hard to exclude usages *to* an external ns. 

Suppose one of my project files looks like this:

```clojure
(defn foo [a] a)
```
And I change it to this:
```clojure
(defn foo [a] (+ 1 a))
```

I've added a usage of `clojure.core/+`. It doesn't matter whether I filter the analysis to be one file or only project files, that internal file will still have that **usage** of `clojure.core/+`. 

Later, when we're looking through the project analysis (which we've filtered to be only internal files), we discover that `clojure.core/+` isn't **defined** in any internal files, and so we don't need to notify anything.

But the comment is asking whether there's an easy way to exclude that usage early, before we look for any definitions. If we could do that, then we wouldn't need to check whether it's defined internally.

If we could quickly get a set of the internal namespaces, we could exclude external usages like `clojure.core/+` before looking through the project definitions. But, I think to find the internal namespaces you'd have to loop through the project analysis again. I haven't benchmarked it, but I expect that would make the whole process slower. Our set of usages contains 1 element. We'd loop through 140000 elements, finding internal namespaces, just so we could get our set of usages down to 0 elements.

In this particular example, that may save us some time. When we have 0 usages, we don't need to look for any definitions, and so we can skip that step. But if we had say, one usage of an internal var and one of an external var, the difference is whether we loop over 140000 elements once, looking for two definitions, or loop over 140000 elements once looking for internal namespaces, and then loop over 140000 elements again, looking for one definition. The former is probably faster.

Anyway, I think I've convinced myself this would be an unhelpful micro-optimization, so I'll remove the comment.

Reply:
Yeah, makes sense, I think we are feeling the need for a way to easily(and fast) know all the project or external analysis on multiple places of the project without filtering all the time whole analysis. Maybe we should research a little bit later how to achieve that, maybe have a pos process or something that would classify in the analysis db a `:internal` and `:external`.

Reply:
👍 To elaborate... there are two parts. First, it'd be nice to quickly know which files are `:internal` or `:external`, to speed up `q/filter-project-analysis`. Second, it'd be nice to know which namespaces are `:internal` or `:external`, to speed up filtering of usages. There's overlap between those two things, but they're slightly different.

```clojure
(def db
  {:external-files [{:file/name "..."
                     :file/uri "..."
                     :file/contents-hash "..." ;; to skip analysis of unchanged external files at init
                     :file/namespaces ['clojure.core]}]
   :internal-files [,,,]
   :analysis {"... file name ..." [,,,]}})

(defn filter-project-analysis [db]
  (let [internal-filenames (map :file/name (:internal-files db))]
    (select-keys (:analysis db) internal-filenames)))

(defn internal-usages [db]
  (let [internal-nses (->> (:internal-files db)
                           (mapcat :file/namespaces)
                           (into #{}))]
    (->> (:analysis db)
         (filter #(identical? :var-usages (:bucket %)))
         (filter #(contains? internal-nses (:to %))))))
```

Reply:
Yes, makes sense, what makes me afraid of creating such data structure is the need to update/invalidate that as analysis change long the time, how we would update? would be fast?

Reply:
Yep, I have the same worries, which is part of why I haven't tried it. A good first step would be to add `db/replace-analysis`. That would let us centralize the logic for maintaining `:analysis` and the file metadata caches. If we decided to change the format, say from a vector of internal files to a hashmap keyed by filename, then we'd only have to update a few places, instead of hunting for everywhere we do a `swap!` on `:analysis`. And we'd only need to get the update/invalidate logic correct in one place. In the future, if we decide we want other caches, for example on bucket, we'd have a central place to do that.

My intuition is that a file metadata cache would be fast, even if implemented pretty naively. Except at startup, you usually only get new analysis of one or a small handful of files. So, deciding whether they're internal or external, finding their namespaces, and inserting them into the cache would all be quick, even if for the cache you use an inconvenient data structure like a vector.

Following this idea a bit further, probably the `db` namespace should be responsible for getting analysis out of the db too, whether that's the full analysis, project analysis, single file analysis, or something else. Then the `queries` namespace could operate on vectors of elements, instead of sometimes that and sometimes a hashmap of filenames to elements.

```clojure
(into []
      (comp db/project-analysis-xf
            (q/filter-bucket-xf :var-usages))
      @db)

;; or

(into []
      (comp (db/file-analysis-xf filename)
            (q/filter-bucket-xf :var-usages))
      @db)
```

Reply:
>  And we'd only need to get the update/invalidate logic correct in one place. In the future, if we decide we want other caches, for example on bucket, we'd have a central place to do that.

I liked it.

> Following this idea a bit further, probably the db namespace should be responsible for getting analysis out of the db too, whether that's the full analysis, project analysis, single file analysis, or something else. Then the queries namespace could operate on vectors of elements, instead of sometimes that and sometimes a hashmap of filenames to elements.

sounds valid too, maybe it could bring other benefits this metadata concept, like avoiding to lint already linted files, that sha idea